File: objc/JukeboxMessed.m
#import<stdio.h>
#import<Foundation/Foundation.h>

/*
 * ジュークボックスエミュレータ
 * 仕様
 * - ジュークボックスは複数の楽曲データを所持している
 * - 一つの楽曲データには、アーティスト名、アルバム名、タイトル、曲の長さが含まれている
 * - ユーザが入力するコマンドに応じて以下の操作を行う
 *   1. 曲の再生
 *   2. プレイリストの自動生成
 *   3. プレイリストの曲飛ばし
 *   4. プレイリストへ曲を追加
 *   5. プレイリストから曲を削除
 *   6. 曲飛ばし方法の変更(順送り、ランダム)
 *   7. 現在の曲情報の表示
 *   8. 楽曲データ一覧の表示
 *   9. アルバム情報の表示
 */

@interface JukeboxMessed : NSObject
{
	NSMutableArray* _artists;
  NSMutableArray* _cdNames;
  NSMutableArray* _songNames;
  NSMutableArray* _songLengthList;
  
  NSMutableArray* _playlist;
  NSString* _selector;
}

- (id)initWithArtists:(NSMutableArray*)artists withCdNames:(NSMutableArray*)cdNames
        withSongNames:(NSMutableArray*)songNames withSongLengthList:(NSMutableArray*)songLengthList;
- (void)activateAuto:(NSArray*)commands;


@property (nonatomic, retain) NSMutableArray* _artists;
@property (nonatomic, retain) NSMutableArray* _cdNames;
@property (nonatomic, retain) NSMutableArray* _songNames;
@property (nonatomic, retain) NSMutableArray* _songLengthList;
@property (nonatomic, retain) NSMutableArray* _playlist;
@property (nonatomic, retain) NSString* _selector;

@end

@implementation JukeboxMessed
- (id)initWithArtists:(NSMutableArray*)artists withCdNames:(NSMutableArray*)cdNames
        withSongNames:(NSMutableArray*)songNames withSongLengthList:(NSMutableArray*)songLengthList
{
	self = [super init];
  _artists = artists;
  _cdNames = cdNames;
  _songNames = songNames;
  _songLengthList = songLengthList;
  
  _playlist = [[NSMutableArray alloc] init];
  _selector = @"normal";
  srand(time(NULL));
	return self;
}

- (void)activateAuto:(NSArray *)commands
{
  for (NSString* cmd in commands) {
    printf("コマンド> %s\n",[cmd UTF8String]);
    NSArray* words = [cmd componentsSeparatedByString:@" "];

    if ([[words objectAtIndex:0] isEqualToString:@"play"]) {
      printf("%s: ", [[_songNames objectAtIndex:[[_playlist objectAtIndex:0] intValue]] UTF8String]);
      for (int i=0 ; i<[[_songLengthList objectAtIndex:0] intValue] ; i++) {
        printf("♪");
      }
      puts("");
    } else if ([[words objectAtIndex:0] isEqualToString:@"next"]) {
      if ([_selector isEqualToString:@"reverse"]) {
        [_playlist exchangeObjectAtIndex:0 withObjectAtIndex:[_playlist count]-1];
        [_playlist removeObjectAtIndex:[_playlist count]-1];
      } else if ([_selector isEqualToString:@"random"]) {
        int target = 1 + (rand() % ((int)[_playlist count]-1));
        [_playlist exchangeObjectAtIndex:1 withObjectAtIndex:target];
        [_playlist removeObjectAtIndex:0];
      } else {
        [_playlist removeObjectAtIndex:0];
      }
    } else if ([[words objectAtIndex:0] isEqualToString:@"selector"]) {
      _selector = [words objectAtIndex:1];
    } else if ([[words objectAtIndex:0] isEqualToString:@"shuffle"]) {
      _playlist = [NSMutableArray array];
      int size = [words count] > 1
                  ? [[words objectAtIndex:1] intValue]
                  : (int)[_songNames count];

      for (int i=0 ; i<size ; i++) {
        [_playlist addObject:[NSNumber numberWithInt:(rand() % (int)[_songNames count])]];
      }
    } else if ([[words objectAtIndex:0] isEqualToString:@"info"]) {
      printf("アーティスト : %s\n", [[_artists objectAtIndex:[[_playlist objectAtIndex:0] intValue]] UTF8String]);
      printf("アルバム : %s\n", [[_cdNames objectAtIndex:[[_playlist objectAtIndex:0] intValue]] UTF8String]);
      printf("タイトル : %s\n", [[_songNames objectAtIndex:[[_playlist objectAtIndex:0] intValue]] UTF8String]);
      printf("長さ : %d\n", [[_songLengthList objectAtIndex:[[_playlist objectAtIndex:0] intValue]] intValue]);
    } else if ([[words objectAtIndex:0] isEqualToString:@"playlist"]) {
      for (NSNumber* i in _playlist) {
        printf("No. %d: ", [i intValue]);
        printf("アーティスト=%s, ", [[_artists objectAtIndex:[i intValue]] UTF8String]);
        printf("アルバム=%s, ", [[_cdNames objectAtIndex:[i intValue]] UTF8String]);
        printf("タイトル=%s, ", [[_songNames objectAtIndex:[i intValue]] UTF8String]);
        printf("長さ=%d\n", [[_songLengthList objectAtIndex:[i intValue]] intValue]);
      }
    } else if ([[words objectAtIndex:0] isEqualToString:@"list"]) {
      if ([words count] > 1) {
        for (int i=1 ; i<[words count] ; i++) {
          for (int j=0 ; j<[_songNames count] ; j++) {
            if ([[_cdNames objectAtIndex:j] isEqualToString:[words objectAtIndex:i]]) {
              printf("%d: %s\n", j, [[_songNames objectAtIndex:j] UTF8String]);
            }
          }
        }
      } else {
        for (int i=0 ; i<[_songNames count] ; i++) {
          printf("%d: %s\n", i, [[_songNames objectAtIndex:i] UTF8String]);
        }
      }
    } else if ([[words objectAtIndex:0] isEqualToString:@"add"]) {
      [_playlist addObject:[NSNumber numberWithInt:[[words objectAtIndex:1] intValue]]];
    } else if ([[words objectAtIndex:0] isEqualToString:@"remove"]) {
      [_playlist removeObjectAtIndex:[[words objectAtIndex:1] intValue]];
    } else if ([[words objectAtIndex:0] isEqualToString:@"cd"]) {
      printf("現在のアルバム: %s\n", [[_cdNames objectAtIndex:[[_playlist objectAtIndex:[[words objectAtIndex:1] intValue]] intValue]] UTF8String]);
    } else {
      printf("not supported\n");
    }
  }
}


@synthesize _artists;
@synthesize _cdNames;
@synthesize _songNames;
@synthesize _songLengthList;
@synthesize _playlist;
@synthesize _selector;
@end

int main(void) {
	NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
  NSMutableArray* artists = [NSMutableArray array];
  NSMutableArray* cdNames = [NSMutableArray array];
  NSMutableArray* songNames = [NSMutableArray array];
  NSMutableArray* songLengthList = [NSMutableArray array];

	[artists addObject:@"SiM"];
	[cdNames addObject:@"EViLS"];
  [songNames addObject:@"Blah Blah Blah"];
  [songLengthList addObject:[NSNumber numberWithInt:8]];
  
  [artists addObject:@"SiM"];
	[cdNames addObject:@"EViLS"];
  [songNames addObject:@"Same Sky"];
  [songLengthList addObject:[NSNumber numberWithInt:3]];

  [artists addObject:@"SiM"];
	[cdNames addObject:@"EViLS"];
  [songNames addObject:@"faith"];
  [songLengthList addObject:[NSNumber numberWithInt:9]];
  
  [artists addObject:@"Fact"];
	[cdNames addObject:@"burundanga"];
  [songNames addObject:@"FOSS"];
  [songLengthList addObject:[NSNumber numberWithInt:3]];
  
  [artists addObject:@"Fact"];
	[cdNames addObject:@"burundanga"];
  [songNames addObject:@"1000 miles"];
  [songLengthList addObject:[NSNumber numberWithInt:13]];
  
  [artists addObject:@"Fact"];
	[cdNames addObject:@"burundanga"];
  [songNames addObject:@"pink rolex"];
  [songLengthList addObject:[NSNumber numberWithInt:5]];


	JukeboxMessed* jukebox = [[JukeboxMessed alloc ]
                            initWithArtists:artists
                            withCdNames:cdNames
                            withSongNames:songNames
                            withSongLengthList:songLengthList];
  [jukebox activateAuto:[NSArray arrayWithObjects:@"list",@"list EViLS",@"shuffle",@"playlist",@"info",
                         @"play",nil]];


  [pool release];
}

File: php/JukeboxMessed.php
<html>
<head>
<title>ジュークボックスシミュレータ</title>
</head>
<body>
<pre>
<?php
class JukeboxMessed {
	function __construct($artists, $cdNames, $songNames, $songLengthList) {
		$this->artists = $artists;
		$this->cdNames = $cdNames;
		$this->songNames = $songNames;
		$this->songLengthList = $songLengthList;

		$this->playlist = array();
		$this->selector = 'normal';
	}

	public function activateAuto($commands) {
		foreach ($commands as $command) {
			echo "コマンド> $command\n";
			$words = split(' +', $command);
			switch ($words[0]) {
			case 'play':
				echo $this->songNames[$this->playlist[0]], ': ';
				for ($i=0 ; $i<$this->songLengthList[$this->playlist[0]] ; $i++) {
					echo '♪';
				}
				echo "\n";
				break;
			case 'next':
				switch ($this->selector) {
				case 'reverse':
					$this->playlist[0] = $this->playlist[count($this->playlist)-1];
					array_splice($this->playlist, count($this->playlist)-1, 1);
					break;
				case 'random':
					$tmp_i = rand(1, count($this->playlist)-1);
					if ($tmp_i != 1) {
						$tmp_p = $this->playlist[1];
						$this->playlist[1] = $this->playlist[$tmp_i];
						$this->playlist[$tmp_i] = $tmp_p;
					}
					array_splice($this->playlist, 0, 1);
					break;
				case 'normal':
				default:
					array_splice($this->playlist, 0, 1);
				}
				break;
			case "selector":
				$this->selector = $words[1];
				break;
			case "shuffle":
				$this->playlist = array();
				$size = count($words)==1 ? count($this->songNames) : $words[1];
				for ($i=0 ; $i<$size ; $i++) {
					$this->playlist[] = rand(0, count($this->songNames)-1);
				}
				break;
			case "info":
				echo "アーティスト: ", $this->artists[$this->playlist[0]], "\n";
				echo "アルバム: ", $this->cdNames[$this->playlist[0]], "\n";
				echo "タイトル: ", $this->songNames[$this->playlist[0]], "\n";
				echo "長さ: ", $this->songLengthList[$this->playlist[0]], "\n";
				break;
			case "playlist":
				for ($i=0 ; $i<count($this->playlist) ; $i++) {
					echo 'No. ', $this->playlist[$i],
						': アーティスト=', $this->artists[$this->playlist[$i]],
						', タイトル=', $this->songNames[$this->playlist[$i]],
						', 長さ=', $this->songLengthList[$this->playlist[$i]], "\n";
				}
				break;
			case "list":
				if (count($words) > 1) {
					for ($i=1 ; $i<count($words) ; $i++) {
						for ($j=0 ; $j<count($this->songNames) ; $j++) {
							if ($this->cdNames[$j] == $words[$i]) {
								echo $j, ': ', $this->songNames[$j], "\n";
							}
						}
					}
				} else {
					for ($i=0 ; $i<count($this->songNames) ; $i++) {
						echo $i, ': ', $this->songNames[$i], "\n";
					}
				}
				break;
			case "add":
				$this->playlist[] = $words[1];
				break;
			case "remove":
				array_splice($this->playlist, $words[1], 1);
				break;
			case "cd":
				echo '現在のアルバム: ', $this->cdNames[$this->playlist[0]], "\n";
				break;
			default:
				echo "コマンド \"$command\" はサポートされていません。";
			}
		}
	}
}


$artists = array();
$cdNames = array();
$songNames = array();
$songLengthList = array();


$artists[] = 'SiM';
$cdNames[] = 'EViLS';
$songNames[] = 'Blah Blah Blah';
$songLengthList[] = 8;

$artists[] = 'SiM';
$cdNames[] = 'EViLS';
$songNames[] = 'Same Sky';
$songLengthList[] = 3;

$artists[] = 'SiM';
$cdNames[] = 'EViLS';
$songNames[] = 'faith';
$songLengthList[] = 9;

$artists[] = 'Fact';
$cdNames[] = 'burundanga';
$songNames[] = 'FOSS';
$songLengthList[] = 3;

$artists[] = 'Fact';
$cdNames[] = 'burundanga';
$songNames[] = '1000 miles';
$songLengthList[] = 13;

$artists[] = 'Fact';
$cdNames[] = 'burundanga';
$songNames[] = 'pink rolex';
$songLengthList[] = 5;


$jukebox = new JukeboxMessed($artists, $cdNames, $songNames, $songLengthList);
$jukebox->activateAuto(array('list'
	, 'list burundanga EViLS'
	, 'add 1'
	, 'add 3'
	, 'playlist'
	, 'remove 0'
	, 'playlist'
	, 'info'
	, 'play'
	, 'shuffle'
	, 'playlist'
	, 'next'
	, 'playlist'
	, 'selector reverse'
	, 'next'
	, 'playlist'
	, 'selector random'
	, 'next'
	, 'playlist'
	, 'cd'
));

?>
</pre>
</body>
</html>
